src/app/delegation_backend/src/itn_uptime_analyzer/uptime.go

```diff
@@ -0,0 +1,346 @@
+						if err != nil {
+							log.Fatalf("Error parsing time: %v\n", err)
+						}
+
+						if (lastSubmissionTimeString != "") && (currentSubmissionTime.After(lastSubmissionTime.Add(time.Duration(syncPeriod-5) * time.Minute))) && (currentSubmissionTime.Before(lastSubmissionTime.Add(time.Duration(syncPeriod+5) * time.Minute))) {
+							uptimeToday = append(uptimeToday, true)
+							lastSubmissionTimeString = submissionDataToday.CreatedAt
+						} else if lastSubmissionTimeString == "" {
+							uptimeToday = append(uptimeToday, true)
+							lastSubmissionTimeString = submissionDataToday.CreatedAt
+						}
+
+					}
+				} else {
+					if (identity["public-key"] == submissionDataToday.Submitter.String()) && (identity["public-ip"] == submissionDataToday.RemoteAddr) {
+						if lastSubmissionTimeString != "" {
+							lastSubmissionTime, err = time.Parse(time.RFC3339, lastSubmissionTimeString)
+							if err != nil {
+								log.Fatalf("Error parsing time: %v\n", err)
+							}
+						}
+
+						currentSubmissionTime, err := time.Parse(time.RFC3339, submissionDataToday.CreatedAt)
+						if err != nil {
+							log.Fatalf("Error parsing time: %v\n", err)
+						}
+
+						if (lastSubmissionTimeString != "") && (currentSubmissionTime.After(lastSubmissionTime.Add(time.Duration(syncPeriod-5) * time.Minute))) && (currentSubmissionTime.Before(lastSubmissionTime.Add(time.Duration(syncPeriod+5) * time.Minute))) {
+							uptimeToday = append(uptimeToday, true)
+							lastSubmissionTimeString = submissionDataToday.CreatedAt
+						} else if lastSubmissionTimeString == "" {
+							uptimeToday = append(uptimeToday, true)
+							lastSubmissionTimeString = submissionDataToday.CreatedAt
+						}
+
+					}
+				}
+			}
+		}
+
+		// Convert everything into percentage
+		uptimePercent := (float64(len(uptimeToday)) / float64(numberOfSubmissionsNeeded)) * 100
+		if uptimePercent > 100.00 {
+			uptimePercent = 100.00
+		}
+		identity["uptime"] = fmt.Sprintf("%.2f%%", uptimePercent)
+	}
+}
+
+// This function does the same as the one above only that it calculates the difference between the time elapsed today and the sync period
+// and checks the previous days bucket for that amount of time
+func (identity Identity) GetUptimeOfTwoDays(config AppConfig, sheet *sheets.Service, ctx dg.AwsContext, log *logging.ZapEventLogger, sheetTitle string, currentTime time.Time, syncPeriod int, executionInterval int) {
+	currentDate := currentTime.Format("2006-01-02")
+	lastExecutionTime := GetLastExecutionTime(config, sheet, log, sheetTitle, currentTime, executionInterval)
+
+	yesterdaysDate := lastExecutionTime.Format("2006-01-02")
+
+	numberOfSubmissionsNeeded := (60 / syncPeriod) * executionInterval
+
+	prefixToday := strings.Join([]string{ctx.Prefix, "submissions", currentDate}, "/")
+	prefixYesterday := strings.Join([]string{ctx.Prefix, "submissions", yesterdaysDate}, "/")
+
+	inputToday := &s3.ListObjectsV2Input{
```

Comment:
I don't like that we traverse all submissions for each identity. It's very expensive.
My suggestion is to implement one loop and have a `map[string][]*Identity` from pubkey to list of identity objects with such public key.

Then when traversing the submissions, we could extract pubkey from the name of submission (no need to read the data) and if it's in the pubkey map, then read submission and process it for related identities.